### PR TITLE
api: add forbidden to api/user-settings

### PR DESCRIPTION
**Pull request by marharyta**
# Checklist for author

Remove items that aren't applicable, check items that are done.

## Reviewability
- [x] Link to issue (#2531)
- [ ] Note if PR is on top of other PR
- [ ] Note if related change in rems-deploy repo
- [ ] Consider adding screenshots for ease of review

## Backwards compatibility
- [x] API is backwards compatible or completely new
- [x] Events are backwards compatible
- [x] Config is backwards compatible

## Documentation
- [ ] Update changelog if necessary - TODO
- [x] API is documented and shows up in Swagger UI
- [ ] Components are added to guide page
- [ ] Update docs/ (if applicable)
- [ ] Update manual/ (if applicable)
- [ ] ADR for major architectural decisions or experiments
- [ ] New config options in config-defaults.edn

## Testing
- [ ] Complex logic is unit tested
- [ ] Valuable features are integration / browser / acceptance tested automatically

## Follow-up
- [ ] New tasks are created for pending or remaining tasks

### CODE REVIEW

**Comment by marharyta** [comment]
@opqdonut @Macroz adding the forbidden path triggers infinite ridirect loop

**Comment by marharyta** [comment]
After a brief discussion with  @opqdonut we came to a conclusion that solution proposed by @Macroz would be the most practical. I agree, it is good to have fallback anyways.

**Comment by marharyta** [comment]
Screenshot 2021-01-29 at 9 53 01

**Comment by marharyta** [comment]
Screenshot 2021-02-01 at 12 48 45
This changes triggers error on login page load due to the fact that we through error since the user is not logged in.

**Comment by Macroz** [comment]
> This changes triggers error on login page load due to the fact that we through error since the user is not logged in.

Well with these changes we explicitly still show the error message while we should eat the error message completely for this case (of not logged-in).

See this part:
```
:error-handler (flash-message/default-error-handler :top "Fetch user settings")}
```

**Comment by marharyta** [comment]
The latest changes seem to reflect the desired outcome: 
- no wrapper over the fetch and no error message on login page
- what I believe could be done further is checking the log-in status and then fetching with logged in status

**Comment by marharyta** [comment]
update: after a brief conversation with @Macroz we decided to create a separate dispatcher event for the error case of the fetch call, I am currently working on implementing it

**Comment by marharyta** [comment]
Can you confirm @Macroz this is the right direction?

**Comment by marharyta** [comment]
I know we had a catch up today, yet I just wanna be extra sure

**Comment by marharyta** [comment]
Also, this is not ready yet, still have to:

- [ ] double check why ```/default-error-handler```  doesn't throw error in UI
- [ ] run tests
- [ ] figure out how to mimic fetch-user settings fail when logged in

**Comment by marharyta** [comment]
@opqdonut any tips on how to test?

**Comment by opqdonut** [comment]
An API-level test-case should be easy to add to `rems.api.test-user-settings`. That's something we should do.

What I'm less sure about is browser testing. I guess we could add a test case for navigating to `/profile` without logging in and checking that we get forwarded to the front page. However I'm not sure this feature is important enough to warrant a slow browser test case.

Some unit tests for our `fetch` util would be nice, but we don't have any scaffolding for that kind of testing yet, and maybe t's not worth it to set that up right now.

File: src/cljs/rems/user_settings.cljs
Comment (Macroz) [inline src/cljs/rems/user_settings.cljs:103]: Good news, @opqdonut and I discussed this and Joel noticed that since we only use `fetch-user-settings!` for the initial fetch (and not for user settings page) we can skip these dispatch events and just ignore everything when the response code is 401. We can skip the logged-in part since this doesn't need really need the complication.

File: src/cljs/rems/user_settings.cljs
Comment (Macroz) [inline src/cljs/rems/user_settings.cljs:107]: You could do a simple check like `(if (= 401 (:status response)) ... ...)` here.

File: src/cljs/rems/user_settings.cljs
Comment (marharyta) [reply src/cljs/rems/user_settings.cljs]: I do not see the login part as a complication, but rather as a logical extension

File: src/cljs/rems/user_settings.cljs
Comment (opqdonut) [inline src/cljs/rems/user_settings.cljs:86]: So is there still a TODO here? The if below looks a bit weird. Why not show the status code always or never?

File: src/cljs/rems/user_settings.cljs
Comment (opqdonut) [inline src/cljs/rems/user_settings.cljs:96]: What's the difference between `-fail` and `-error`? Could they be combined?

You could also consider just inlining the event(s) into the `:error-handler`. We have something like that in `rems.application/save-attachment` for example.

File: src/cljs/rems/user_settings.cljs
Comment (opqdonut) [inline src/cljs/rems/user_settings.cljs:106]: no need to merge multiple single-key maps: just use a literal map

File: src/cljs/rems/util.cljs
Comment (opqdonut) [inline src/cljs/rems/util.cljs:91]: Could document this option in the docstring. The current docstring is a bit weird, but a short mention about this option would fit I think.

**Review by opqdonut — CHANGES_REQUESTED** [review]

File: src/cljs/rems/user_settings.cljs
Comment (marharyta) [reply src/cljs/rems/user_settings.cljs:94]: I think I just got confused with the initial 401 suggestion

File: src/cljs/rems/util.cljs
Comment (marharyta) [reply src/cljs/rems/util.cljs:91]: did that

File: src/cljs/rems/user_settings.cljs
Comment (opqdonut) [inline src/cljs/rems/user_settings.cljs:84]: now that this event is just a oneliner, I'd rather get rid of it (and inline it into `:loaded-user-settings-fail`)

File: src/cljs/rems/user_settings.cljs
Comment (opqdonut) [inline src/cljs/rems/user_settings.cljs:97]: Oh right, now I remember that discussion about 401. I think the point was that it would be nice to show an error when

- the user is not logged in
- *but* the settings API fails with some other code than 401

In other words, the error should _only be hidden if_:

- the user is not logged in
- and the error was 401

File: src/clj/rems/api/user_settings.clj
Comment (opqdonut) [inline src/clj/rems/api/user_settings.clj:14]: you could add a not logged in case to `rems.api.test-user-settings`

**Review by opqdonut — CHANGES_REQUESTED** [review]
sorry, I got some ideas I could've also posted in the previous review

File: src/cljs/rems/user_settings.cljs
Comment (Macroz) [reply src/cljs/rems/user_settings.cljs]: A logical extension that is not useful is basically unnecessary logic and code that should be simplified i.e. not exist?

File: test/clj/rems/api/test_user_settings.clj
Comment (Macroz) [inline test/clj/rems/api/test_user_settings.clj:24]: I think there is some confusion here. This test both checks the return value for `unauthorized` and calls a function called `assert-unauthorized`. I'm sure both should not be needed.

File: src/cljs/rems/user_settings.cljs
Comment (Macroz) [inline src/cljs/rems/user_settings.cljs:92]: How about we simplify this code to something like this?
```
   {:db db
    :dispatch (when (not= 401 (:status user-settings-error)))
                (flash-message/show-default-error! :top (str "Fetch user settings")))}
```

But then we could also inline this below like @opqdonut suggested because there is no necessary reference to `db`.

Btw. I think the description string should be localized here.

File: src/cljs/rems/user_settings.cljs
Comment (Macroz) [inline src/cljs/rems/user_settings.cljs:98]: Can just inline the code here.

File: src/cljs/rems/user_settings.cljs
Comment (Macroz) [inline src/cljs/rems/user_settings.cljs:96]: I'm not sure why this was renamed to `:simple-error-handler?` but I think it's backwards. Isn't the normal i.e. simple case that you wrap the default error handler? And this would then be the ... `:wrap-default-error-handler? false` or even `:custom-error-handler true` case. Do you see what I mean?

File: src/cljs/rems/util.cljs
Comment (Macroz) [inline src/cljs/rems/util.cljs:79]: The confusion about the parameter name is evident in this sentence too, isn't it? This doesn't really say what it does or what the wrapper it's talking about is. One has to read the code.

File: test/clj/rems/api/testing.clj
Comment (Macroz) [inline test/clj/rems/api/testing.clj:168]: This is named `assert-` yet it does no such thing wrt. to unauthorized. Copying some fields from the response makes no sense to me.

**Review by Macroz — CHANGES_REQUESTED** [review]

File: src/cljs/rems/user_settings.cljs
Comment (marharyta) [reply src/cljs/rems/user_settings.cljs]: If you are talking about ```:flash-message-error```  then it has been addressed

File: src/cljs/rems/user_settings.cljs
Comment (marharyta) [reply src/cljs/rems/user_settings.cljs:92]: Yes, I agree, yet, just to note, we did not have it localized before.

File: src/cljs/rems/user_settings.cljs
Comment (marharyta) [reply src/cljs/rems/user_settings.cljs:98]: I would leave it like this, if you do not have a strong opinion on it.

File: src/cljs/rems/util.cljs
Comment (marharyta) [reply src/cljs/rems/util.cljs:79]: Yes, I agree. The thing is due to so many iterations it is hard to keep the naming clear, there were a lot of discussions about this fix, first from @opqdonut then we had a call with you, then you and Joel discussed it between yourselves ... I will take a look at it with some fresh eyes and try to read through it and then maybe the naming change will be more apparent to me.

File: test/clj/rems/api/testing.clj
Comment (marharyta) [reply test/clj/rems/api/testing.clj:168]: hmm, I was referencing the logic of the other functions in the file, but I guess they are sort of all over the place. 
Ill double check. thanks!

File: src/cljs/rems/user_settings.cljs
Comment (marharyta) [reply src/cljs/rems/user_settings.cljs:96]: same here, need to go through the ticket again to think of the better naming

File: test/clj/rems/api/test_user_settings.clj
Comment (marharyta) [reply test/clj/rems/api/test_user_settings.clj:24]: hmm, I guess I should create something like ```read-unauthorized```

File: src/cljs/rems/user_settings.cljs
Comment (Macroz) [reply src/cljs/rems/user_settings.cljs:92]: One more note. ACtually from an `fx` you don't need to return the `db` if you don't touch it.

File: test/clj/rems/api/test_user_settings.clj
Comment (Macroz) [reply test/clj/rems/api/test_user_settings.clj:24]: Why does it need to do it? The fields are available in the response object as they are, aren't they? And you can check the status if you so wish. No helper required.

File: test/clj/rems/api/testing.clj
Comment (Macroz) [reply test/clj/rems/api/testing.clj:168]: They serve a specific purpose, i.e. so that the threaded code works nicely and you get a clean error if the assertion was not met. This helper is not needed for the test at all, I think.

File: src/cljs/rems/util.cljs
Comment (Macroz) [reply src/cljs/rems/util.cljs:79]: I think the most important and also the most exciting part of a PR is once you have the solution implemented. Then you refactor (the diff and) the solution until it is actually nice. I.e. "first make it work, then make it right, ...". Only at this stage can you see what the naming actually should be, what is unnecessary, etc.

File: src/cljs/rems/user_settings.cljs
Comment (marharyta) [reply src/cljs/rems/user_settings.cljs:98]: I think it looks rather consistent this way.

File: src/cljs/rems/user_settings.cljs
Comment (marharyta) [reply src/cljs/rems/user_settings.cljs:92]: Is this a suggestion or a requirement?

Comment (Macroz) [reply src/cljs/rems/user_settings.cljs:92]: I'll let @opqdonut decide. I would definitely do this refactor, it's like a one minute thing to fix the code.

File: src/cljs/rems/user_settings.cljs
Comment (Macroz) [reply src/cljs/rems/user_settings.cljs:98]: Yeah that is right, the simplification is elsewhere.

File: test/clj/rems/api/testing.clj
Comment (marharyta) [reply test/clj/rems/api/testing.clj]: I guess I can resolve it since I deleted this code?

File: src/cljs/rems/user_settings.cljs
Comment (marharyta) [reply src/cljs/rems/user_settings.cljs]: Do you think naming is more compelling now?

File: src/cljs/rems/user_settings.cljs
Comment (marharyta) [reply src/cljs/rems/user_settings.cljs:92]: I would like to clarify: are you talking about the localization or the one liner?

File: src/cljs/rems/user_settings.cljs
Comment (Macroz) [reply src/cljs/rems/user_settings.cljs]: Actually I don't. I mean if I am passing `:error-handler XXX` to me it means I'm also passing a custom error handler so I don't understand why I need to pass another parameter called `:custom-error-handler` to do something else? It affects the wrapping so I would've liked to see that word in it. :shrug:

File: src/cljs/rems/user_settings.cljs
Comment (opqdonut) [reply src/cljs/rems/user_settings.cljs:92]: Please simplify this code a bit. The repeated `:db db` and nested ifs make the logic look a lot more complicated than it is, at least to me. At least remove the `:db` keys, and also consider combining the conditions into one if.

File: test/clj/rems/api/testing.clj
Comment (opqdonut) [inline test/clj/rems/api/testing.clj:149]: This is seems like a useful little helper, but please fix the name. It doesn't have anything to do with unauthorized does it? How about a name like `status-and-body` or `read-body-and-status`?

File: src/cljs/rems/util.cljs
Comment (opqdonut) [inline src/cljs/rems/util.cljs:79]: Picking a good name for this parameter is hard. Let's stick with this for now.

**Review by opqdonut — APPROVED** [review]
Good to go after fixing the two minor points: the :loaded-user-settings-fail implementation, and the naming of read-unauthorized-body.

File: test/clj/rems/api/testing.clj
Comment (marharyta) [reply test/clj/rems/api/testing.clj:149]: Yes, that is true, I was naming it based on purpose, not the action, my bad

File: src/cljs/rems/user_settings.cljs
Comment (opqdonut) [inline src/cljs/rems/user_settings.cljs:87]: an if without an else case could be a when

File: src/cljs/rems/user_settings.cljs
Comment (opqdonut) [inline src/cljs/rems/user_settings.cljs:88]: you could just skip the `:db db` altogether: you only need to invoke the `:db` effect if you change the db!

**Comment by marharyta** [comment]
ping @opqdonut

File: src/cljs/rems/user_settings.cljs
Comment (opqdonut) [inline src/cljs/rems/user_settings.cljs:89]: could combine this when with the above if

**Review by opqdonut — APPROVED** [review]
feel free to merge

some code golfing suggestions included

File: src/cljs/rems/user_settings.cljs
Comment (marharyta) [reply src/cljs/rems/user_settings.cljs:88]: interesting, I saw somewhere that db has to be returned every time in re-frame, but I guess I either misunderstood or that was an old resource

Comment (opqdonut) [reply src/cljs/rems/user_settings.cljs:88]: yes, a `reg-event-db` needs to return the new state of the db, but a `reg-event-fx` returns a map of effects, and all events are optional!

Comment (marharyta) [reply src/cljs/rems/user_settings.cljs]: ```events are optional!``` like the enthusiasm! :)

Comment (marharyta) [reply src/cljs/rems/user_settings.cljs]: ok, I think I did address those comments

File: src/cljs/rems/user_settings.cljs
Comment (marharyta) [reply src/cljs/rems/user_settings.cljs]: addressed

File: src/cljs/rems/user_settings.cljs
Comment (marharyta) [reply src/cljs/rems/user_settings.cljs]: yeah

File: test/clj/rems/api/testing.clj
Comment (marharyta) [reply test/clj/rems/api/testing.clj]: addressed

**Review by opqdonut — APPROVED** [review]

**Pull request merged**
